srt-protocol/src/protocol/pending_connection/listen.rs

```diff
@@ -182,42 +225,105 @@ impl Listen {
         }
     }
 
-    fn handle_control_packets(
+    fn request_access(
         &mut self,
-        control: ControlPacket,
-        from: SocketAddr,
-        now: Instant,
-        acceptor: &mut impl StreamAcceptor,
+        remote: SocketAddr,
+        local_socket_id: SocketId,
+        timestamp: TimeStamp,
+        state: ConclusionWaitState,
+        shake: HandshakeControlInfo,
+        incoming: HsV5Info,
     ) -> ConnectionResult {
-        match (self.state.clone(), control.control_type) {
-            (InductionWait, ControlTypes::Handshake(shake)) => {
-                self.wait_for_induction(from, control.timestamp, shake, now)
-            }
-            (ConclusionWait(state), ControlTypes::Handshake(shake)) => {
-                self.wait_for_conclusion(from, control.timestamp, state, shake, now, acceptor)
-            }
-            (InductionWait, control_type) | (ConclusionWait(_), control_type) => {
-                NotHandled(ConnectError::HandshakeExpected(control_type))
-            }
-        }
+        // TODO: handle StreamId parsing error
+        let sid = incoming.sid.clone().unwrap_or_default().into_bytes();
```

Comment:
Hmm, this seems to discard the difference between no StreamId an empty StreamId, which is a difference we do expose to the AccessController API (considering it takes an Option<&str>)

Reply:
An empty StreamId, which is essentially a String with a max length constraint, is legitimate, just like an empty String is legit. There's no benefit to adding a second layer of "nothing" semantics to the type. I would change the AccesssController API to just use &str, actually the intent was to have it use StreamId, just, again, this is a rough draft. I might even just simplify the AccessController API to just a Fn(AccessControlRequest) -> AccessControlResponse, and maybe Box<dyn> it to reduce monomorphisms.

Reply:
An empty StreamId, which is essentially a String with a max length constraint, is legitimate, just like an empty String is legit. There's no benefit to adding a second layer of "nothing" semantics to the type. I would change the AccesssController API to just use &str, actually the intent was to have it use StreamId, just, again, this is a rough draft. I might even just simplify the AccessController API to just a Fn(AccessControlRequest) -> AccessControlResponse, and maybe Box<dyn> it to reduce monomorphisms.

Reply:
But on the packet level, there is a differentiation between empty streamid and no streamid, and is also exposed on the sender side.

Reply:
Ok, you persuaded me, I"ll change all the APIs to reflect this.